_posts/2021-09-07-what-javascript-tests-could-learn-from-rspec.md

```diff
@@ -0,0 +1,275 @@
+---
+layout: epic
+title: "What JavaScript Tests Could Learn From RSpec"
+date: 2021-09-10
+categories: [testing, rspec, jest]
+author: steve-hicks
+---
+
+TODO: rel="canonical" to the original on my blog
+
+When I started at Artsy a few years ago, I'd never written a line of Ruby. I feel at home with JavaScript — it's
+been my buddy since I started my career over 20 years ago. I've written enough tests in JavaScript that I sometimes
+feel like I can write them in my sleep (as long as they don't involve async React events 😅).
+
+Most of the code I write at Artsy is still JavaScript, but now I write some Ruby code too, and I've written enough
+RSpec tests that I'm starting to form opinions about what I think they should look like.
+
+My most recent work has been JavaScript again. I've been writing Jest tests against one of our React apps. But
+rather than reaching for the testing patterns I'd become accustomed to over my years of JavaScripting, I'm finding
+that something's missing in my Jest tests! My experiences with RSpec have me longing for two features in Jest:
+
+<!-- more -->
+
+1. `context` blocks
+2. `let` blocks
+
+## 1. `context` blocks
+
+A [`context` block](https://relishapp.com/rspec/rspec-core/v/2-11/docs/example-groups/basic-structure-describe-it)
+in an RSpec test is, as I understand it, literally the same thing as a `describe` block. Like it's just an alias of
+`describe`. What's the point, you ask?
+
+The difference is in, well, context. Well-organized RSpec tests use `describe` to describe what's being
+tested...and `context` to describe scenarios of the thing being tested.
+
+For example, if I wanted to test the `multiply` method of a `Calculator` class, I might write some test scenarios
+that look like this:
+
+```rb
+describe "Calculator" do
+  describe ".multiply" do
+    context "when the first value is negative" do
+      context "when the second value is negative" do
+        it "returns a positive number" do
+        end
+      end
+      context "when the second value is positive" do
+        it "returns a negative number" do
+        end
+      end
+    end
+  end
+end
+```
+
+See the difference in those test cases between `describe` and `context`? The way I think about it is: if the
+statement coming after my `describe`/`context` describes a pre-condition for the test, it's a `context`; otherwise
+it's a `describe`.
+
+`context` wouldn't be hard to implement in JavaScript — I'd bet there are test frameworks that have it. It'd just
```

Comment:
I'd back adding this alias to jest

Reply:
I'll look into suggesting it!

Reply:
@orta is the best way to start this conversation to open a ["feature request" issue](https://github.com/facebook/jest/issues?q=context+label%3A%22%3Arocket%3A+Feature+Request%22)?